src/main/java/com/google/sps/data/algorithm/TspSolver.java

```diff
@@ -0,0 +1,414 @@
+   *     index 1: (int) --     cases 1, 2, 4: amount of time added by action
+   *                           case 3: Integer.MAX_VALUE
+   * Details per case:
+   * Case 1: add the amount of time till it opens plus time spent (one hour)
+   * Case 2 & 4: add travel time plus time spent (one hour)
+   * Case 3: add Integer.MAX_VALUE
+   */
+  private List<Object> checkIsOpenSetTime(LocalTime currentTime, int i,
+      int currPos) {
+    // list of length 2 to return
+    List<Object> output = new ArrayList<>();
+
+    boolean isOpen; // boolean to be set and added at index 0
+    int timeAdded; // int to be set and added at index 1
+      if (openHours.get(i) != null) {
+        LocalTime open = openHours.get(i).open.time;
+        LocalTime close = openHours.get(i).close.time;
+        
+        // edge case where close time is next day at 0:00
+        if (close.compareTo(LocalTime.parse("00:00")) == 0) {
+          close = LocalTime.of(23, 59);
+        } else {
+          close.minusMinutes(HOUR);
+        }  
+
+        // set time after travelTime
+        LocalTime afterDrive = currentTime.plusMinutes(timeMatrix[currPos][i]);
+
+        /**
+         * check if arrival time is before open time, during,
+         * or after closing time
+         */ 
+        if (afterDrive.compareTo(open) < 0) {
+          isOpen = true;
+          timeAdded = (int) afterDrive.until(open, ChronoUnit.MINUTES) + 
+              timeMatrix[currPos][i] + HOUR;
+        } else if ((afterDrive.compareTo(open) >= 0) && 
+            (afterDrive.compareTo(close) <= 0)) {
+          isOpen = true;
+          timeAdded = timeMatrix[currPos][i] + HOUR;
+        } else {
+          isOpen = false;
+          timeAdded = Integer.MAX_VALUE;
+        }
+      } else {
+        isOpen = true;
+        timeAdded = timeMatrix[currPos][i] + HOUR;
+      }
+
+      // add calculated values and return output
+      output.add(isOpen);
+      output.add(timeAdded);
+      return output;
+  }
+
+  /**
+   * Create time Matrix using DirectionsRequests (DirectionsAPI)
+   * 
+   * @param center central location
+   * @param pois list of pois other than central location
+   */
+  private void createTimeMatrix(String center, List<String> pois) 
+      throws IOException {
+
+    // create allPois to hold `center` and all other `pois`
+    List<String> allPois = new ArrayList<>();
+    allPois.add(center);
+    allPois.addAll(pois);
+
+    // create timeMatrix and fill diagonal with 0's
+    timeMatrix = new int[allPois.size()][allPois.size()];
+    for (int i = 0; i < allPois.size(); i++) {
+      timeMatrix[i][i] = 0;
+    }
+
+    // fill (i, j) and (j, i) indices with travel time from i to j
+    for (int i = 0; i < allPois.size(); i++) {
+      for (int j = i; j < allPois.size(); j++) {
+        int travelTime = getTravelTimeMins(allPois.get(i), allPois.get(j));
+        timeMatrix[i][j] = travelTime;
+        timeMatrix[j][i] = travelTime;
```

Comment:
Nice! I like you use the symetric nature of travel times to only query i->j then fill in both the [i][j] and [j][i] locations in the matrix.

I'm not sure but the DistanceMatrixApi might let you get all the values you need in a single request which would improve the performance here even more: https://github.com/googlemaps/google-maps-services-java/blob/master/src/main/java/com/google/maps/DistanceMatrixApi.java

Reply:
I understand that I could use this however this has calculates all times between locations (from A to B, B to A, and from A to A). For all pairs of locations. Requesting from the server may take longer but it will also take restructuring of the `createTimeMatrix()` method so i will keep it as is.

(current function makes `n^2 / 2 - n` requests. Given that n will never  be too large since its the number of locations per day I believe it's reasonable to keep as is)